cli/coverage.rs

```diff
@@ -0,0 +1,250 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+
+use crate::colors;
+use crate::file_fetcher::SourceFile;
+use crate::inspector::DenoInspector;
+use deno_core::v8;
+use deno_core::ErrBox;
+use futures::channel::mpsc;
+use serde::Deserialize;
+use std::mem::MaybeUninit;
+use std::ops::Deref;
+use std::ops::DerefMut;
+use std::ptr;
+
+pub struct CoverageCollector {
+  v8_channel: v8::inspector::ChannelBase,
+  v8_session: v8::UniqueRef<v8::inspector::V8InspectorSession>,
+  sender: mpsc::Sender<String>,
+  receiver: mpsc::Receiver<String>,
+}
+
+impl Deref for CoverageCollector {
+  type Target = v8::inspector::V8InspectorSession;
+  fn deref(&self) -> &Self::Target {
+    &self.v8_session
+  }
+}
+
+impl DerefMut for CoverageCollector {
+  fn deref_mut(&mut self) -> &mut Self::Target {
+    &mut self.v8_session
+  }
+}
+
+impl v8::inspector::ChannelImpl for CoverageCollector {
+  fn base(&self) -> &v8::inspector::ChannelBase {
+    &self.v8_channel
+  }
+
+  fn base_mut(&mut self) -> &mut v8::inspector::ChannelBase {
+    &mut self.v8_channel
+  }
+
+  fn send_response(
+    &mut self,
+    _call_id: i32,
+    message: v8::UniquePtr<v8::inspector::StringBuffer>,
+  ) {
+    let message = message.unwrap().string().to_string();
+    self.sender.try_send(message).unwrap();
+  }
+
+  fn send_notification(
+    &mut self,
+    _message: v8::UniquePtr<v8::inspector::StringBuffer>,
+  ) {
+  }
+
+  fn flush_protocol_notifications(&mut self) {}
+}
+
+impl CoverageCollector {
+  const CONTEXT_GROUP_ID: i32 = 1;
+
+  pub fn new(inspector_ptr: *mut DenoInspector) -> Box<Self> {
+    new_box_with(move |self_ptr| {
+      let v8_channel = v8::inspector::ChannelBase::new::<Self>();
+      let v8_session = unsafe { &mut *inspector_ptr }.connect(
+        Self::CONTEXT_GROUP_ID,
+        unsafe { &mut *self_ptr },
+        v8::inspector::StringView::empty(),
+      );
+
+      let (sender, receiver) = mpsc::channel::<String>(1);
+
+      Self {
+        v8_channel,
+        v8_session,
+        sender,
+        receiver,
+      }
+    })
+  }
+
+  async fn dispatch(&mut self, message: String) -> Result<String, ErrBox> {
+    let message = v8::inspector::StringView::from(message.as_bytes());
+    self.v8_session.dispatch_protocol_message(message);
+
+    let response = self.receiver.try_next()?;
```

Comment:
Not actually an async chain because of this, intended to use a thread initially but it wasn't necessary.
Should we make this all sync or poll here?

Reply:
Prefer to use future here